diff --git a/grub-core/osdep/linux/getroot.c b/grub-core/osdep/linux/getroot.c
index 001b818..e950edd 100644
--- a/grub-core/osdep/linux/getroot.c
+++ b/grub-core/osdep/linux/getroot.c
@@ -725,10 +725,13 @@ grub_util_part_to_disk (const char *os_dev, struct stat *st,
 {
   char *path;
 
-  if (! S_ISBLK (st->st_mode))
+  if (! getenv("GRUB_PERMIT_NON_BLOCK_DEVICES"))
     {
-      *is_part = 0;
-      return xstrdup (os_dev);
+      if (! S_ISBLK (st->st_mode))
+        {
+          *is_part = 0;
+          return xstrdup (os_dev);
+        }
     }
 
   path = xmalloc (PATH_MAX);
